components/google_sign_in/browser/google_sign_in_throttle.cc

```diff
@@ -0,0 +1,150 @@
+    delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+  }
+  // In case of ASK, we need to be careful because delegate may be deleted
+  return;
+}
+
+std::unique_ptr<blink::URLLoaderThrottle>
+GoogleSignInThrottle::MaybeCreateThrottleFor(
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter& wc_getter,
+    HostContentSettingsMap* content_settings) {
+  if (!IsGoogleSignInFeatureEnabled()) {
+    return nullptr;
+  }
+
+  if (request.resource_type !=
+      static_cast<int>(blink::mojom::ResourceType::kMainFrame)) {
+    return nullptr;
+  }
+
+  const auto request_url = request.url;
+  const auto request_initiator_url =
+      request.request_initiator.value_or(url::Origin()).GetURL();
+
+  // We might not have a request_initiator_url if the request is coming from
+  // a top-level navigation and it's a redirect. We still want to create the
+  // throttle.
+  if (!request_url.SchemeIsHTTPOrHTTPS()) {
+    return nullptr;
+  }
+
+  return std::make_unique<GoogleSignInThrottle>(
+      wc_getter, request_url,
+      base::WrapRefCounted<HostContentSettingsMap>(content_settings));
+}
+
+GoogleSignInThrottle::~GoogleSignInThrottle() = default;
+
+void GoogleSignInThrottle::DetachFromCurrentSequence() {}
+
+void OnPermissionDeny(URLLoaderThrottle::Delegate* delegate) {
+  delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+  return;
+}
+
+void HandleRequest(bool* defer,
+                   const GURL& request_url,
+                   const GURL& request_initiator_url,
+                   const content::WebContents::Getter& contents_getter,
+                   scoped_refptr<HostContentSettingsMap> content_settings,
+                   URLLoaderThrottle::Delegate* delegate) {
+  auto* contents = contents_getter.Run();
+  if (!contents)
+    return;
+
+  if (!request_initiator_url.is_valid() || !request_url.is_valid() ||
+      !IsGoogleAuthRelatedRequest(request_url, request_initiator_url)) {
+    return;
+  }
+
+  // Check kGoogleLoginControlType pref and cancel request if false
+  // NOTE: This means that if the kGoogleLoginControlType permission in
+  // is turned off, all requests to kGoogleAuthPattern and kFirebaseUrlPattern
+  // will be disallowed
+  PrefService* prefs =
+      user_prefs::UserPrefs::Get(contents->GetBrowserContext());
+
+  if (!IsGoogleSignInPrefEnabled(prefs)) {
+    delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+    return;
+  }
+
+  GetPermissionAndMaybeCreatePrompt(
+      contents, request_initiator_url, defer,
+      base::BindOnce(&OnPermissionRequestStatus,
+                     contents->GetController().GetPendingEntry(), contents,
+                     request_initiator_url, content_settings, delegate),
+      base::BindOnce(&OnPermissionDeny, delegate));
+}
+
+void GoogleSignInThrottle::WillStartRequest(network::ResourceRequest* request,
+                                            bool* defer) {
+  const GURL request_url = request->url;
```

Comment:
just pass this directly to the function below.

Reply:
I think it's slightly easier to read this way, at least for me